Add hooks to `RunWorkflow`.

samussiah opened the pull request:
## Overview
Adds a `LoadData` hook at the beginning of `RunWorkflow` and a `SaveData` hook at the end.

## Test Notes/Sample Code
Simplest use case that hits `clindata` and writes to .csv, also captured in `inst/examples/WorkflowIO.R`:

```
load_all()

LoadData <- function(lWorkflow, lConfig) {
    purrr::imap(
        lWorkflow$spec,
        ~ {
            input <- lConfig$Domains[[ .y ]]

            if (is.data.frame(input)) {
                data <- input
            } else if (is.function(input)) {
                data <- input()
            } else if (is.character(input)) {
                data <- read.csv(input)
            } else {
                cli::cli_abort("Invalid data source: {input}.")
            }

            return(ApplySpec(data, .x))
        }
    )
}

SaveData <- function(lWorkflow, lConfig) {
    domain <- paste0(lWorkflow$meta$Type, '_', lWorkflow$meta$ID)
    cli::cli_alert_info(domain)

    if (exists(domain, lConfig$Domains)) {
        output <- lConfig$Domains[[ domain ]]
        cli::cli_alert_info(output)

        cli::cli_alert_info(
            'Saving output of `lWorkflow` to `{output}`.'
        )

        write.csv(
            lWorkflow$lResult,
            output
        )
    } else {
        cli::cli_alert_info(
            '{domain} not found.'
        )
    }
}

lConfig <- list(
    LoadData = LoadData,
    SaveData = SaveData,
    Domains = list(
        Raw_PD = function() { clindata::ctms_protdev },
        Raw_SUBJ = function() { clindata::rawplus_dm },
        Raw_SITE = function() { clindata::ctms_site },

        Mapped_PD = paste0(tempdir(), '/mapped-pd.csv'),
        Mapped_SUBJ = paste0(tempdir(), '/mapped-subj.csv'),
        Mapped_SITE = paste0(tempdir(), '/mapped-site.csv')
    )
)

lMappedData <- RunWorkflows(
    MakeWorkflowList(c('SUBJ', 'PD', 'SITE')),
    lConfig = lConfig
)
```

## Connected Issues
- Closes #1818

samussiah commented:
Still working on unit tests @lauramaxwell but the functionality is there.

lauramaxwell commented:
> Still working on unit tests @lauramaxwell but the functionality is there.

Nice- took a quick look and everything in the example worked as i would expect. I'll dig in a bit more in the morning, but so far so good! let me know if you want me to look into the tests, etc.

lauramaxwell commented on R/RunWorkflow.R:
```diff
@@ -8,6 +8,9 @@
 #' @param lData `list` A named list of domain-level data frames.
+#' @param lConfig `list` A configuration object with two methods:
+#' - `LoadData`: A function that loads data specified in `lWorkflow$spec`.
+#' - `SaveData`: A function that saves data returned by the last step in `lWorkflow$steps`.
```
it might be helpful to add the `4_WorkflowIO.R` example to the examples in this roxygen header, or just make a reference to the example in this param explanation

samussiah replied:
Added in the latest commit.

samussiah commented:
> looks good! added a comment about adding an example to the roxygen header, but feel free to do with that as you wish.

Fixed the in-memory example and added a configuration-based example.  Unit tests are in as well.

lauramaxwell commented on R/RunWorkflow.R:
```diff
@@ -99,6 +199,27 @@ RunWorkflow <- function(
+        is.function(lConfig$SaveData) &&
+        all(c('lWorkflow', 'lConfig') %in% names(formals(lConfig$SaveData)))
+      ) {
+        cli::cli_h3('Saving data with `lConfig$SaveData`.')
```
@zdz2101 FYI- I'm going to merge this in before the logging PR gets merged, so a few new `cli::` calls will need to be changed over in that PR once this gets merged to dev

lauramaxwell commented:
> > looks good! added a comment about adding an example to the roxygen header, but feel free to do with that as you wish.
> 
> Fixed the in-memory example and added a configuration-based example. Unit tests are in as well.

Great! the new examples are super helpful. i've noted a few new `cli` calls that will need to be adjusted in #1939, but that's not a blocker. Merging this now!
